pkg/cache/v3/resource.go

```diff
@@ -111,57 +117,103 @@ func MarshalResource(resource types.Resource) (types.MarshaledResource, error) {
 	return b.Bytes(), nil
 }
 
+func GetResourceReferences(resources map[string]types.ResourceWithTTL) map[resource.Type]map[string]bool {
+	out := make(map[resource.Type]map[string]bool)
+	GetResourceReferencesByReference(resources, out)
+
+	return out
+}
+
 // GetResourceReferences returns the names for dependent resources (EDS cluster
-// names for CDS, RDS routes names for LDS).
-func GetResourceReferences(resources map[string]types.ResourceWithTTL) map[string]bool {
-	out := make(map[string]bool)
+// names for CDS, RDS/SRDS routes names for LDS).
+func GetResourceReferencesByReference(resources map[string]types.ResourceWithTTL, out map[resource.Type]map[string]bool) {
 	for _, res := range resources {
 		if res.Resource == nil {
 			continue
 		}
 		switch v := res.Resource.(type) {
 		case *endpoint.ClusterLoadAssignment:
-			// no dependencies
+			// No dependencies.
 		case *cluster.Cluster:
-			// for EDS type, use cluster name or ServiceName override
-			switch typ := v.ClusterDiscoveryType.(type) {
-			case *cluster.Cluster_Type:
-				if typ.Type == cluster.Cluster_EDS {
-					if v.EdsClusterConfig != nil && v.EdsClusterConfig.ServiceName != "" {
-						out[v.EdsClusterConfig.ServiceName] = true
-					} else {
-						out[v.Name] = true
-					}
-				}
-			}
+			getClusterReferences(v, out)
```

Comment:
I'm a bit more of a fan of this function returning a map instead of taking an `out` param. That we can we can remove the nil checks and just always assume one will exist since we create it. 

Then here maybe do something like
```go
return getClusterReference(v)
```

Reply:
Yeah it was originally just returning a new param instead of taking a reference param, but there was some inefficiency with it later on where you would have to combine all the maps you made later on. Instead, I opted to just use 1 shared map.

I think I have a better solution which meets both of our concerns though, using what you recommended. (see my long winded comment below)

Reply:
Nvm, I get it.